Add CS5460A power-meter component

balrog-kun opened the pull request:
## Description:

CS5460A is a chip that calculates power, RMS current, voltage, energy
etc. and communicates over SPI.  Easy to buy on a breakout board.

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#944

## Checklist:
  - [x] The code change is tested and works locally.
  - [kinda] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

balrog-kun committed:
Add CS5460A power-meter component

CS5460A is a chip that calculates power, RMS current, voltage, energy
etc. and communicates over SPI.  Easy to buy on a breakout board.

glmnet commented on esphome/components/cs5460a/cs5460a.h:
```diff
@@ -0,0 +1,116 @@
+  bool reset_();
+  void restart_();
+
+  uint32_t samples_{4000};
+  int8_t phase_offset_{0};
+  CS5460APGAGain pga_gain_{CS5460A_PGA_GAIN_10X};
+  float current_gain_{0.001};
+  float voltage_gain_{0.001};
+  bool current_hpf_{true};
+  bool voltage_hpf_{true};
+  float pulse_energy_wh_{10.0};
+  float current_multiplier_;
+  float voltage_multiplier_;
+  float power_multiplier_;
+  sensor::Sensor *current_sensor_{nullptr};
+  sensor::Sensor *voltage_sensor_{nullptr};
+  sensor::Sensor *power_sensor_{nullptr};
+  const char *init_status_{"none"};
+  bool running_{false};
```
can you remove all initializers here? those are all initialized by generated code

balrog-kun replied:
Ok, will fix.

glmnet commented on esphome/components/cs5460a/cs5460a.cpp:
```diff
@@ -0,0 +1,249 @@
+  if (!this->reset_()) {
+    this->disable();
+    ESP_LOGE(TAG, "CS5460A reset failed!");
+    init_status_ = "reset failed";
```
mark component failed here, see component api

glmnet commented on esphome/components/cs5460a/cs5460a.cpp:
```diff
@@ -0,0 +1,249 @@
+}
+
+void CS5460AComponent::loop() {
+  if (!running_)
```
may be not needed is component marked failed

glmnet commented on esphome/components/cs5460a/cs5460a.cpp:
```diff
@@ -0,0 +1,249 @@
+  LOG_SENSOR("  ", "Power", power_sensor_);
+}
+
+void CS5460AComponent::loop() {
```
is this desired to query every loop? usually sensors inherit PolloingComponent so `update_interval`can be used

balrog-kun replied:
This is a good question.  In my use case I don't want to miss any of the values the sensor is reporting because it can watch only one power line at a time and I want to monitor various lines so it takes it a few seconds to go over all the power lines, and I'd like esphome to react quickly to changes.  Others might want to integrate (accumulate) the power values so that they can monitor how much was consumed in a day for example, so they also don't want to miss any values.

But then there may be people who want to publish the values once a minute for example, so they can use the `throttle` filter but the esp32 will be wasting some cycles, so I'm not sure what the best solution is.

I wanted to avoid reading the SPI register on every loop() by calculating when the next sampling period is expected end.  Let's say we read new values at time T0 and we know that the period takes about 1s, so we can skip loop() until maybe T0 + 900ms and then start polling the register.  But I'm not sure if running millis() on every loop is much faster than doing the SPI register access.  What do you think?

glmnet replied:
You mean that you do a query and the thing takes about 900ms to have a sample ready?

If that is the case I would use millis and sleep for a while and then check every 50ms if a response is ready.  Mostly in case other stuff like display etc would like to use spi and leave them time to update quickly. You don't know what other users do and microcontrollers are a compromise in resources. 

Also if this updates HA too quickly you should suggest throttle in docs (haven't read them yet in case you did)

glmnet replied:
Reading millis and skipping spi io is of course much less hassle for the CPU

balrog-kun replied:
Ok, I'll do that.  Maybe even loop() can be empty, and I can do everything in ``set_timeout()`` callbacks.

It's not strictly 900ms, normally the device does continuous calculations and updates the measurements every N cycles.  You can set your own interval but it's important to set it to an integer number of AC cycles (for example for 50 Hz AC current it's better if the time is a multiple of 20ms).

The default is 1 second, so I thought about waking up 100ms  before that (1s - 100ms = 900ms) and start polling the STATUS register until the data is ready.  I'll add a warning about using `throttle`.

balrog-kun commented:
Ok, thanks @glmnet.  Doing all the work in `set_timeout()` callbacks now, `loop()` is empty and everything works.

After each measurement we schedule STATUS register reads (through SPI) at 0.8x the expected time left until next reading but not less than 30ms.  So for example when we expect a new data at T + 1 second, we do an SPI read at:
T + 800ms
T + 960ms
T + 992ms
and then every 30ms until data ready or until 500ms have passed.

Also added runtime reset logic for when the device gets stuck for 15s, and added the usage of `status_momentary_warning()` on non-fatal errors.

balrog-kun commented:
Rebased to master/dev, added default device_class params to the sensor_schema() usages and added the CODEOWNERS tag.

jesserockz committed:
Merge branch 'dev' into cs5460a

jesserockz committed:
Format file

jesserockz committed:
Use async

jesserockz committed:
Remove `sensor` prefix from action

jesserockz committed:
Add enalbe/disable around `restart_` function which is used by the restart action
